contrib/win32/win32compat/ssh-agent/keyagent-request.c

```diff
@@ -209,36 +307,42 @@ static int sign_blob(const struct sshkey *pubkey, u_char ** sig, size_t *siglen,
 	struct sshbuf* tmpbuf = NULL;
 	char *keyblob = NULL;
 	const char *sk_provider = NULL;
+#ifdef ENABLE_PKCS11
+	int is_pkcs11_key = 0;
+#endif /* ENABLE_PKCS11 */
 
 	*sig = NULL;
 	*siglen = 0;
 
-	if ((thumbprint = sshkey_fingerprint(pubkey, SSH_FP_HASH_DEFAULT, SSH_FP_DEFAULT)) == NULL ||
-	    get_user_root(con, &user_root) != 0 ||
-	    RegOpenKeyExW(user_root, SSH_KEYS_ROOT,
-			0, STANDARD_RIGHTS_READ | KEY_QUERY_VALUE | KEY_WOW64_64KEY | KEY_ENUMERATE_SUB_KEYS, &reg) != 0 ||
-	    RegOpenKeyExA(reg, thumbprint, 0,
-			STANDARD_RIGHTS_READ | KEY_QUERY_VALUE | KEY_ENUMERATE_SUB_KEYS | KEY_WOW64_64KEY, &sub) != 0 ||
-	    RegQueryValueExW(sub, NULL, 0, NULL, NULL, &regdatalen) != ERROR_SUCCESS ||
-	    (regdata = malloc(regdatalen)) == NULL ||
-	    RegQueryValueExW(sub, NULL, 0, NULL, regdata, &regdatalen) != ERROR_SUCCESS ||
-	    convert_blob(con, regdata, regdatalen, &keyblob, &keyblob_len, FALSE) != 0 ||
-	    (tmpbuf = sshbuf_from(keyblob, keyblob_len)) == NULL)
-		goto done;
-
+#ifdef ENABLE_PKCS11
+	if ((prikey = lookup_key(pubkey)) == NULL) {
+#endif /* ENABLE_PKCS11 */
+		if ((thumbprint = sshkey_fingerprint(pubkey, SSH_FP_HASH_DEFAULT, SSH_FP_DEFAULT)) == NULL ||
+			get_user_root(con, &user_root) != 0 ||
+			RegOpenKeyExW(user_root, SSH_KEYS_ROOT,
+				0, STANDARD_RIGHTS_READ | KEY_QUERY_VALUE | KEY_WOW64_64KEY | KEY_ENUMERATE_SUB_KEYS, &reg) != 0 ||
+			RegOpenKeyExA(reg, thumbprint, 0,
+				STANDARD_RIGHTS_READ | KEY_QUERY_VALUE | KEY_ENUMERATE_SUB_KEYS | KEY_WOW64_64KEY, &sub) != 0 ||
+			RegQueryValueExW(sub, NULL, 0, NULL, NULL, &regdatalen) != ERROR_SUCCESS ||
+			(regdata = malloc(regdatalen)) == NULL ||
+			RegQueryValueExW(sub, NULL, 0, NULL, regdata, &regdatalen) != ERROR_SUCCESS ||
+			convert_blob(con, regdata, regdatalen, &keyblob, &keyblob_len, FALSE) != 0 ||
+			(tmpbuf = sshbuf_from(keyblob, keyblob_len)) == NULL ||
+			sshkey_private_deserialize(tmpbuf, &prikey) != 0)
```

Comment:
Add a log here. It's very difficult to debug what went wrong

Reply:
Added. Thanks!